### PR TITLE
Capture S3 bucket+key as otel attributes

### LINKED ISSUE #3006: [META 759] Add S3 Bucket and Object Key to S3 instrumentation
See meta issue and spec for the description and details:
- Meta issue: https://github.com/elastic/apm/issues/759
- Spec issue: https://github.com/elastic/apm/issues/760

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Fixes #3006

Captures the following as OTel attributes from [OTel semantic conventions for S3](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/instrumentation/aws-sdk/#s3).
- `aws.s3.bucket`
- `aws.s3.key`
- `aws.s3.copy_source`

The following attributes are not covered
- `aws.s3.upload_id`
- `aws.s3.delete`
- `aws.s3.part_number`

This PR is quite large due to test refactoring.

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] ~~Added an API method or config option? Document in which version this will be introduced~~ n/a
  - [x] ~~I have made corresponding changes to the documentation~~ n/a

### CODE REVIEW

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/main/java/co/elastic/apm/agent/awssdk/v1/helper/S3Helper.java [hunk 1]
```diff
@@ -35,4 +37,36 @@ public static S3Helper getInstance() {
     public S3Helper(Tracer tracer) {
         super(tracer, SdkV1DataSource.getInstance());
     }
+
+    @Nullable
+    @Override
+    protected String getObjectKey(Request<?> request, @Nullable String bucketName) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/main/java/co/elastic/apm/agent/awssdk/v1/helper/S3Helper.java:43]: [for reviewer] with SDK v1 we have to use some manual parsing of an HTTP header, this is not the case with v2.

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/test/java/co/elastic/apm/agent/awssdk/v1/DynamoDbClientIT.java [hunk 2]
```diff
@@ -75,40 +74,63 @@ public void setupClient() {
...
+            .withKeySchema(List.of(
+                new KeySchemaElement("attributeOne", KeyType.HASH),
+                new KeySchemaElement("attributeTwo", KeyType.RANGE)
+            ))
+            .withBillingMode(BillingMode.PAY_PER_REQUEST)))
+            .operationName("CreateTable")
+            .entityName(TABLE_NAME)
+            .action("query")
+            .withSpanAssertions(dbAssert)
+            .execute();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/test/java/co/elastic/apm/agent/awssdk/v1/DynamoDbClientIT.java:91]: [for reviewer] this is mostly automatic refactoring with a "test builder" that allows to combine assertions on the spans in a more declarative way.

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/test/java/co/elastic/apm/agent/awssdk/v1/S3ClientIT.java [hunk 3]
```diff
@@ -53,25 +52,79 @@ public void setupClient() {
...
-        executeTest("ListBuckets", null, () -> s3.listBuckets());
-        executeTest("PutObject", BUCKET_NAME, () -> s3.putObject(BUCKET_NAME, OBJECT_KEY, "This is some Object content"));
-        executeTest("ListObjects", BUCKET_NAME, () -> s3.listObjects(BUCKET_NAME));
-        executeTest("GetObject", BUCKET_NAME, () -> s3.getObject(BUCKET_NAME, OBJECT_KEY));
-        executeTest("CopyObject", NEW_BUCKET_NAME, () -> s3.copyObject(BUCKET_NAME, OBJECT_KEY, NEW_BUCKET_NAME, "new-key"));
-        executeTest("DeleteObject", BUCKET_NAME, () -> {
+        newTest(() -> s3.createBucket(BUCKET_NAME))
+            .operationName("CreateBucket")
+            .entityName(BUCKET_NAME)
+            .otelAttribute("aws.s3.bucket", BUCKET_NAME)
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/test/java/co/elastic/apm/agent/awssdk/v1/S3ClientIT.java:58]: [for reviewer] this tests the expected OTel attribute capture that was added with this PR.

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/S3Helper.java [hunk 4]
```diff
@@ -35,4 +38,22 @@ public static S3Helper getInstance() {
...
+        }
+
+        return value;
+    }
+
+    @Nullable
+    @Override
+    protected String getCopySource(SdkRequest request) {
+        return awsSdkDataSource.getFieldValue(IAwsSdkDataSource.COPY_SOURCE_FIELD, request);
+    }
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/S3Helper.java:58]: [for reviewer] with SDK v2 we don't need to parse anything from the request, just capture it.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java [hunk 5]
```diff
@@ -336,6 +336,11 @@ public T withSync(boolean sync) {
         return thiz();
     }
 
+    @Override
+    public T withOtelAttribute(String key, @Nullable Object value) {
+        return this.withOTelAttribute(key, value);
```
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java]: That was a leftover from IDE refactoring, I've not removed it.

**Review by JonasKunz — APPROVED** [review]

**Pull request merged**